### PR TITLE
Prioritize remote address from `X-Forwarded-For` header when possible

### PR DESCRIPTION
**Pull request by tylerbenson**
and similar for port.

### CODE REVIEW

**Comment by richardstartin** [comment]
This may seem like the continuation of a running joke, but how on earth did this get to be a 100 file, 900 line diff? Shouldn't we be changing a few lines of code in a decorator somewhere and fixing up the tests? I thought the point of the decorator and test class hierarchies was to abstract away implementation details like this so they are easy to make?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 1]
```diff
@@ -20,13 +20,18 @@
 
   private static final BitSet SERVER_ERROR_STATUSES = Config.get().getHttpServerErrorStatuses();
 
+  public static final String FORWARDED_FOR_HEADER = "x-forwarded-for";
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:23]: nit: these are only actually used in one place, can't they just be inline literals?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 2]
```diff
@@ -77,7 +87,15 @@ public AgentSpan onConnection(final AgentSpan span, final CONNECTION connection)
         }
       }
 
-      setPeerPort(span, peerPort(connection));
+      if (forwardedPort != null) {
+        try {
+          setPeerPort(span, Integer.parseInt(forwardedPort));
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:92]: Do we really need to parse the port? We will send it to the agent as a string anyway.

File: dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java [hunk 3]
```diff
@@ -20,8 +20,7 @@ public static AgentScope createSpan(final HttpRequest request) {
     span.setMeasured(true);
 
     DECORATE.afterStart(span);
-    DECORATE.onConnection(span, request);
-    DECORATE.onRequest(span, request);
+    DECORATE.onRequest(span, request, request);
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java:23]: Looks like an improvement but seems unrelated - what's the rationale?

**Review by richardstartin — COMMENTED** [review]
I read the whole thing and understand why the diff is so large - most of the changes are in test code - but it would have been smaller with typed spans. I would be happy to approve once the Finatra tests are fixed.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 4]
```diff
@@ -20,13 +20,18 @@
...
+  private static final String FORWARDED_PORT_HEADER = "x-forwarded-port";
+
   // Assigned here to avoid repeat boxing and cache lookup.
   public static final Integer _500 = HTTP_STATUSES.get(500);
 
   protected abstract String method(REQUEST request);
 
   protected abstract URIDataAdapter url(REQUEST request);
 
+  protected abstract String header(REQUEST request, String header);
```
Comment (bantonsson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:33]: I guess that adding this type of generic header access is ok for this use case, and as long as it's not used to look up a large number of headers.

File: dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java (hunk 3 above)
Comment (bantonsson) [reply dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java:23]: I"m fine with collapsing these, since they are almost exclusively called right after each other.

**Review by bantonsson — COMMENTED** [review]
I'm wondering if this type of common header extraction would have been better served by an `HttpCodec.Extractor` that adds the tags, and then a check if the port/host was present before setting them in the decorator?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java (hunk 1 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:23]: If you'd prefer it as an inline literal, I can do that.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:92]: Good point.  I'll make this change.

File: dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/akka-http-10.0/src/main/java/datadog/trace/instrumentation/akkahttp/DatadogWrapperHelper.java:23]: The rationale with doing it here is the port is accessed via the connection where the headers are accessed via the request, so in order to evaluate this properly I needed to have access to both.

**Comment by tylerbenson** [comment]
I think I've applied all the requested changes.  Please review again.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 5]
```diff
@@ -63,21 +64,22 @@ public AgentSpan onRequest(final AgentSpan span, final REQUEST request) {
...
 
-  public AgentSpan onConnection(final AgentSpan span, final CONNECTION connection) {
     if (connection != null) {
-      final String ip = peerHostIP(connection);
-      if (ip != null) {
-        if (ip.indexOf(':') > 0) {
-          span.setTag(Tags.PEER_HOST_IPV6, ip);
-        } else {
-          span.setTag(Tags.PEER_HOST_IPV4, ip);
+      if (span.getTag(Tags.PEER_HOST_IPV4) == null && span.getTag(Tags.PEER_HOST_IPV6) == null) {
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:69]: Why do we need these checks now? This will tank in high volume services.

Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:69]: Because we only want to overwrite the tags if they have not been set by the extractor, which happens before we get here. That being said, the logic could be switched around, so we only do the checks if we get back an `ip` or `port`, and maybe only check the `IPV4` or `IPV6` and not both.

Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:69]: But the value we're doing this work to capture is never interesting unless you have a direct client to server connection, which I don't think anybody does these days.

Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:69]: Are you proposing that we completely remove this code? Another way to minimize the cost of the checks would be to pass along the extracted context here as well and then access the fields in the `TagContext`.

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/B3HttpCodec.java [hunk 6]
```diff
@@ -97,6 +101,10 @@ public boolean accept(String key, String value) {
           classification = SPAN_ID;
         } else if (SAMPLING_PRIORITY_KEY.equalsIgnoreCase(key)) {
           classification = SAMPLING_PRIORITY;
+        } else if (FORWARDED_FOR_KEY.equalsIgnoreCase(key)) {
+          classification = FORWARDED_FOR;
+        } else if (FORWARDED_PORT_KEY.equalsIgnoreCase(key)) {
+          classification = FORWARDED_PORT;
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/B3HttpCodec.java:107]: I was wondering if it wouldn't be possible to extract this common code into the `ContextInterpreter` in a method that would allow us to exit early if the `x-forwarded-*` header was handled. Something along the lines of:

```
  protected boolean handledForwarding(String key, String value) {
    if (FORWARDED_FOR_KEY.equalsIgnoreCase(key)) {
      String firstValue = firstHeaderValue(value);
      if (null != firstValue) {
        forwardedFor = firstValue;
      }
      return true;
    }
    if (FORWARDED_PORT_KEY.equalsIgnoreCase(key)) {
      String firstValue = firstHeaderValue(value);
      if (null != firstValue) {
        forwardedPort = firstValue;
      }
      return true;
    }
    return false;
  }
```

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:69]: Yes, I'm suggesting its utility is limited to keeping some tests passing. The IP address we report is currently the address of the proxy in all deployments except where clients interact directly with a server instance. I doubt we have users who do that. 

I think it's useful to keep the costs of the things we do in mind, so we can balance the decisions to collect data. This is a profile of a traced Tomcat request. The bit in the square below is our overhead. 
Screenshot 2021-03-09 at 12 13 42

From left to right, that's made of:

the scope manager activating the span

Screenshot 2021-03-09 at 12 21 57

starting the span 

Screenshot 2021-03-09 at 12 22 10


the scope manager closing the scope 

Screenshot 2021-03-09 at 12 22 15


extracting the context from the headers

Screenshot 2021-03-09 at 12 22 24

the decorator getting the port in `onConnection` and `onRequest` (setting tags/tag interception, getting the host, getting the path, building the url), 

Screenshot 2021-03-09 at 12 22 31

finishing the span (which is dominated by publishing metrics now), 

Screenshot 2021-03-09 at 12 22 37

some actual Tomcat work involved in finishing the request, 

Screenshot 2021-03-09 at 12 22 42

`afterStart` (setting tags/tag interception, more setting tags).

Screenshot 2021-03-09 at 12 22 48


So;
* the cost of interacting with the tags is high, it's better to model data with fields in objects. 
* when we collect data like the port and IP address here, it usually looks innocuous, but often triggers lazy work, which wouldn't happen without installing the tracer. Does the data we're collecting from the request really add any value in this case?

Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java]: I've applied the suggested change.

**Review by richardstartin — APPROVED** [review]

**Review by bantonsson — APPROVED** [review]

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/B3HttpCodec.java (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-core/src/main/java/datadog/trace/core/propagation/B3HttpCodec.java:107]: The implementations have enough differences that I think it would require refactoring to make this work well. I'm going to punt on this for now.
